File: e2e/native/java/src/test/java/e2e/GeneratedTest.java
package e2e;

import com.example.contracts.NativeCases;
import com.example.contracts.EdgeCases;
import com.example.contracts.Token;
import com.example.contracts.Vault;
import java.math.BigInteger;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.web3j.abi.TypeEncoder;
import org.web3j.abi.datatypes.DynamicBytes;
import org.web3j.abi.datatypes.DynamicArray;
import org.web3j.abi.datatypes.generated.Bytes32;
import org.web3j.abi.datatypes.generated.Uint256;
import org.web3j.protocol.core.DefaultBlockParameterName;
import org.web3j.protocol.core.methods.response.Log;
import static org.junit.jupiter.api.Assertions.*;

class GeneratedTest {
    @Test void offlineCallsResultsErrorsAndEvents() {
        String owner = "0x0000000000000000000000000000000000000001";
        assertEquals(Token.BALANCE_OF_SELECTOR, Token.encodeBalanceOf(owner).substring(0, 10));
        assertEquals(BigInteger.TEN, Token.decodeBalanceOfResult("0x" + TypeEncoder.encode(new Uint256(BigInteger.TEN))));
        assertEquals(Token.TRANSFER_SELECTOR, Token.encodeTransfer(owner, BigInteger.TEN).substring(0, 10));
        assertThrows(UnsupportedOperationException.class, () -> NativeCases.encodeStore(
            new NativeCases.Data(new DynamicBytes(new byte[]{1})), java.util.Collections.nCopies(40, BigInteger.ONE)));
        String error = Vault.INSUFFICIENT_SHARES_ERROR_SELECTOR + TypeEncoder.encode(new Uint256(BigInteger.TEN)) + TypeEncoder.encode(new Uint256(BigInteger.ONE));
        assertEquals(BigInteger.TEN, Vault.decodeInsufficientSharesError(error).available());
        Log log = new Log();
        log.setTopics(List.of(Token.TRANSFER_EVENT_TOPIC, "0x" + TypeEncoder.encode(new org.web3j.abi.datatypes.Address(owner)), "0x" + TypeEncoder.encode(new org.web3j.abi.datatypes.Address(owner))));
        log.setData("0x" + TypeEncoder.encode(new Uint256(BigInteger.TEN)));
        assertEquals(BigInteger.TEN, Token.decodeTransferEvent(log).amount());
        assertEquals(3, Token.filterTransferEvent(owner, DefaultBlockParameterName.EARLIEST, DefaultBlockParameterName.LATEST, owner, null).getTopics().size());
    }

    @Test void nestedTupleAndArrayDecode() {
        String target = "0x0000000000000000000000000000000000000001";
        var call = new NativeCases.Call3Value(target, true, BigInteger.TEN, new DynamicBytes(new byte[]{1, 2, 3}));
        var grid = new NativeCases.Grid(List.of(List.of(BigInteger.ONE, BigInteger.TWO), List.of()), List.of(true, false), List.of(new Bytes32(new byte[32])));
        var batch = new NativeCases.Batch(List.of(call), grid, BigInteger.valueOf(3000), BigInteger.valueOf(-5));
        String output = "0x" + TypeEncoder.encode(new Uint256(32)) + TypeEncoder.encode(batch);
        var decoded = NativeCases.decodeBatchValue(output);
        assertEquals(BigInteger.TEN, decoded.calls.get(0).value_);
        assertEquals(List.of(BigInteger.ONE, BigInteger.TWO), decoded.grid.rows.get(0));
        assertEquals(BigInteger.valueOf(3000), decoded.fee);

        var row = new DynamicArray<>(Uint256.class, List.of(new Uint256(BigInteger.ONE), new Uint256(BigInteger.TEN)));
        var matrix = new DynamicArray<>(DynamicArray.class, List.of(row));
        String matrixOutput = "0x" + TypeEncoder.encode(new Uint256(32)) + TypeEncoder.encode(matrix);
        assertEquals(List.of(List.of(BigInteger.ONE, BigInteger.TEN)), EdgeCases.decodeNestedArrayResult(matrixOutput));

        String hash = "0x" + "ab".repeat(32);
        Log log = new Log();
        log.setTopics(List.of(NativeCases.INDEXED_REFERENCES_EVENT_TOPIC, hash, hash, hash));
        log.setData("0x");
        var event = NativeCases.decodeIndexedReferencesEvent(log);
        assertEquals(hash, event.label());
        assertEquals(hash, event.payload());
        assertEquals(hash, event.values());
    }
}

File: e2e/native/kotlin/src/test/kotlin/GeneratedTest.kt
import com.example.contracts.NativeCases
import com.example.contracts.TupleCases
import com.example.contracts.Vault
import com.example.contracts.EdgeCases
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import org.web3j.abi.FunctionEncoder
import org.web3j.abi.datatypes.DynamicBytes
import org.web3j.abi.datatypes.Function
import org.web3j.abi.datatypes.generated.Bytes32
import org.web3j.abi.datatypes.generated.Uint256
import org.web3j.abi.datatypes.Address
import org.web3j.abi.datatypes.DynamicArray
import org.web3j.protocol.core.DefaultBlockParameterName
import org.web3j.protocol.core.methods.response.Log
import kotlin.test.assertFails
import kotlin.test.assertFailsWith

class GeneratedTest {
    @Test
    fun packedMarketDecodesFromFunctionOutput() {
        val market = NativeCases.PackedMarket(Bytes32(ByteArray(32) { 1 }), Bytes32(ByteArray(32) { 2 }), BigInteger.TEN)
        val encoded = org.web3j.abi.TypeEncoder.encode(market)
        val decoded = org.web3j.abi.FunctionReturnDecoder.decode(
            encoded,
            Function("getMarket", emptyList(), listOf(org.web3j.abi.TypeReference.create(NativeCases.PackedMarket::class.java))).outputParameters,
        ).single() as NativeCases.PackedMarket
        assertEquals(market, decoded)
    }

    @Test
    fun arrayTupleDecodesFromFunctionOutput() {
        val amounts = NativeCases.Amounts(listOf(BigInteger.ONE, BigInteger.TEN))
        val encoded = org.web3j.abi.TypeEncoder.encode(amounts)
        // A dynamic return value starts with an offset to its tuple payload.
        val output = org.web3j.abi.TypeEncoder.encode(org.web3j.abi.datatypes.generated.Uint256(32)) + encoded
        val decoded = org.web3j.abi.FunctionReturnDecoder.decode(
            output,
            Function("getAmounts", emptyList(), listOf(org.web3j.abi.TypeReference.create(NativeCases.Amounts::class.java))).outputParameters,
        ).single() as NativeCases.Amounts
        assertEquals(amounts, decoded)
    }

    private fun batch(fee: Long) = NativeCases.Batch(
        calls = listOf(
            NativeCases.Call3Value(
                target = "0x0000000000000000000000000000000000000001",
                allowFailure = true,
                value_ = BigInteger.TEN,
                callData = DynamicBytes(byteArrayOf(1, 2, 3)),
            ),
        ),
        grid = NativeCases.Grid(
            rows = listOf(listOf(BigInteger.ONE, BigInteger.TWO), listOf()),
            flags = listOf(true, false),
            tags = listOf(Bytes32(ByteArray(32) { 7 })),
        ),
        fee = BigInteger.valueOf(fee),
        delta = BigInteger.valueOf(-5),
    )

    @Test
    fun structsEncodeWithTheGeneratedSelector() {
        val encoded = FunctionEncoder.encode(Function("aggregate", listOf(batch(3000)), emptyList()))
        assertEquals(NativeCases.AGGREGATE_SELECTOR, encoded.substring(0, 10))

        val deposit = FunctionEncoder.encode(
            Function(
                "deposit",
                listOf(TupleCases.TupleAccountPosition("0x0000000000000000000000000000000000000002")),
                emptyList(),
            ),
        )
        assertEquals(TupleCases.DEPOSIT_0_SELECTOR, deposit.substring(0, 10))
    }

    @Test
    fun valuesCompareByContent() {
        assertEquals(batch(3000), batch(3000))
        assertEquals(batch(3000).hashCode(), batch(3000).hashCode())
        assertNotEquals(batch(3000), batch(500))
    }

    @Test
    fun abiIsEmbedded() {
        assert(NativeCases.JSON.startsWith("[{"))
    }

    @Test
    fun typedWrapperEncodesAndDecodesOffline() {
        val owner = "0x0000000000000000000000000000000000000001"
        val call = NativeCases.encodeBalanceOf(owner)
        assertEquals(NativeCases.BALANCE_OF_SELECTOR, call.substring(0, 10))
        assertEquals(BigInteger.TEN, NativeCases.decodeBalanceOfResult("0x" + org.web3j.abi.TypeEncoder.encode(Uint256(BigInteger.TEN))))
        val transfer = NativeCases.encodeTransfer(owner, BigInteger.TEN)
        assertEquals(NativeCases.TRANSFER_SELECTOR, transfer.substring(0, 10))
        assertFails { NativeCases.decodeBalanceOfResult("0x") }
        assertFailsWith<UnsupportedOperationException> {
            NativeCases.encodeStore(NativeCases.Data(DynamicBytes(byteArrayOf(1))), List(40) { BigInteger.ONE })
        }

        val recipient = "0x0000000000000000000000000000000000000002"
        assertEquals(Vault.DEPOSIT_0_SELECTOR, Vault.encodeDeposit0(BigInteger.TEN, recipient).substring(0, 10))
        assertEquals(Vault.DEPOSIT_1_SELECTOR, Vault.encodeDeposit1(BigInteger.TEN).substring(0, 10))
        assertEquals(BigInteger.TEN, Vault.TransactionOptions(BigInteger.ONE, BigInteger.TEN, BigInteger.TEN).value)
    }

    @Test
    fun typedWrapperDecodesEventsAndErrors() {
        val from = "0x0000000000000000000000000000000000000001"
        val to = "0x0000000000000000000000000000000000000002"
        val log = Log().apply {
            topics = listOf(
                NativeCases.TRANSFER_EVENT_TOPIC,
                "0x" + org.web3j.abi.TypeEncoder.encode(Address(from)),
                "0x" + org.web3j.abi.TypeEncoder.encode(Address(to)),
            )
            data = "0x" + org.web3j.abi.TypeEncoder.encode(Uint256(BigInteger.TEN))
        }
        assertEquals(NativeCases.TransferEvent(from, to, BigInteger.TEN), NativeCases.decodeTransferEvent(log))
        val filter = NativeCases.filterTransferEvent(from, DefaultBlockParameterName.EARLIEST, DefaultBlockParameterName.LATEST, from, null)
        assertEquals(3, filter.topics.size)
        assertEquals(NativeCases.TRANSFER_EVENT_TOPIC, filter.topics[0].value)

        val encodedError = Vault.INSUFFICIENT_SHARES_ERROR_SELECTOR +
            org.web3j.abi.TypeEncoder.encode(Uint256(BigInteger.TEN)) +
            org.web3j.abi.TypeEncoder.encode(Uint256(BigInteger.ONE))
        assertEquals(BigInteger.TEN, Vault.decodeInsufficientSharesError(encodedError)?.available)
        val dynamicErrorArgs = FunctionEncoder.encodeConstructor(
            listOf(Uint256(BigInteger.ONE), DynamicBytes(byteArrayOf(1, 2, 3))),
        )
        val dynamicError = NativeCases.FAILED_ERROR_SELECTOR + dynamicErrorArgs.removePrefix("0x")
        assertEquals(BigInteger.ONE, NativeCases.decodeFailedError(dynamicError)?.index)
    }

    @Test
    fun typedWrapperDecodesNestedArraysAndTuples() {
        val row = DynamicArray(Uint256::class.java, listOf(Uint256(BigInteger.ONE), Uint256(BigInteger.TEN)))
        val matrix = DynamicArray(DynamicArray::class.java, listOf(row))
        val matrixOutput = "0x" + org.web3j.abi.TypeEncoder.encode(Uint256(32)) + org.web3j.abi.TypeEncoder.encode(matrix)
        assertEquals(listOf(listOf(BigInteger.ONE, BigInteger.TEN)), EdgeCases.decodeNestedArrayResult(matrixOutput))

        val amounts = NativeCases.Amounts(listOf(BigInteger.ONE, BigInteger.TEN))
        val tupleOutput = "0x" + org.web3j.abi.TypeEncoder.encode(Uint256(32)) + org.web3j.abi.TypeEncoder.encode(amounts)
        assertEquals(amounts, NativeCases.decodeGetAmountsResult(tupleOutput))
    }

    @Test
    fun deeplyNestedBatchTupleDecodesWithoutLosingArrayValues() {
        val original = batch(3000)
        val output = "0x" + org.web3j.abi.TypeEncoder.encode(Uint256(32)) + org.web3j.abi.TypeEncoder.encode(original)
        val decoded = NativeCases.decodeBatchValue(output)
        assertEquals(original, decoded)
    }

    @Test
    fun indexedDynamicEventExposesHashes() {
        val hash = "0x" + "ab".repeat(32)
        val log = Log().apply {
            topics = listOf(NativeCases.INDEXED_REFERENCES_EVENT_TOPIC, hash, hash, hash)
            data = "0x"
        }
        val decoded = NativeCases.decodeIndexedReferencesEvent(log)
        assertEquals(hash, decoded?.label)
        assertEquals(hash, decoded?.payload)
        assertEquals(hash, decoded?.values)
    }
}
